Ignore temporary properties during zfs.filesystem.present

asomers opened the pull request:
A temporary property is one that gets set by the mount command's "-o"
option and only applies for the duration of the mount.  Salt's ZFS
module cannot change these, so Salt's zfs state should ignore them.

### What does this PR do?
Ignores temporary ZFS properties during the "filesystem.present" state.

### What issues does this PR fix or reference?
If a temporary mountpoint were set contrary to the defined state, then state.apply would try to correct it, to no effect.  Also, ignoring temporary properties will soon be faster (see https://github.com/openzfs/zfs/pull/11955)

### Previous Behavior
filesystem.present would try in vain to correct a temporary property.  For example "readonly". 

### New Behavior
filesystem.present will ignore temporary properties.

### Merge requirements satisfied?
- [ ] Docs
- [ ] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [XXX] Tests written/updated (tests will be added after #59970 merges)

### Commits signed with GPG?
Yes

asomers committed:
Ignore temporary properties during zfs.filesystem.present

A temporary property is one that gets set by the mount command's "-o"
option and only applies for the duration of the mount.  Salt's ZFS
module cannot change these, so Salt's zfs state should ignore them.

Signed-off-by: Alan Somers <asomers@gmail.com>

asomers commented:
cc @alek-p @xeacott

krionbsd commented:
Thanks, please fix pre-commit and add a test to reflect the change

asomers commented:
I've fixed pre-commit.  But adding tests is blocked on #59970 .

waynew commented:
IIUC the tests will be part of #59970, correct? Or will you be adding more tests for this?

waynew commented:
@asomers I've gone ahead and add the help wanted label to this - if you are able to pick it up, great! Otherwise maybe someone else will be able to get it over the finish line :+1:

asomers commented:
No, the tests cannot be part of 59970.  They would fail if I added them to that PR now.  And if I add them to this PR, they would cause a merge conflict with that one.  59970 must be merged before I can add tests to this PR.

waynew commented:
That makes sense - I noted on the other PR that there are some changes that could be made to make things a bit more idiomatic, but in the interest of unblocking *this* PR I went ahead and approved that. When I review this PR I'll make sure to call those out - two off the top of my head:

- `MagicMock` should *rarely* appear in test code. `patch(..., autospec=True, return_value=...)` or `create_autospec(..., return_value=...)` is more typically appropriate. Autospecs take a little more when running tests, but they make up for it by guarding against changes in underlying function signatures
- `len(x) == 0` / `len(x) > 0` should typically be `if not x:` / `if x:`

asomers commented:
@waynew I rebased this PR and updated all of the ZFS unit tests.  But I didn't add any new ones.  It turns out there really isn't anything that can be unit-tested for this PR, because the interesting part happens within ZFS itself.  If I were to add a test, it would basically be identical to the existing `test_filesystem_present_properties` test.  The only testable change is "The zfs get call should include a `source` argument", and that's covered by the existing tests.

Jsollvander commented:
I'm not sure if the current reviewer is still working with salt? @Ch3LL can you review this PR?

dwoz commented:
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.

asomers commented:
@dwoz @Ch3LL please reopen this PR.  All it needs is a reviewer.

dwoz commented:
@asomers please add a changelog entry too.

asomers commented:
@dwoz changelog file added.

dwoz commented:
This can be re-based and re-opened if it should go into `3008.x`
